### PR TITLE
Code quality improvements

### PR DESCRIPTION
**Pull request by ezyang**
Each commit is self-contained.

### CODE REVIEW

**Comment by apaszke** [comment]
Not sure if it's worth adding back the whole infrastructure for obtaining these Python-side pointers just for this single method. Also, we should consider using pybind since it's part of the repo now

**Comment by ezyang** [comment]
@apaszke It's an investment for the future, in case we decide to add other methods :)

**Comment by apaszke** [comment]
I really think we shouldn't be adding code which doesn't serve any purpose at the moment. Also, once we need it I'd really go with pybind because it makes it a lot simpler. Let's keep these commits around, but not merge them until we actually need them

**Comment by zdevito** [comment]
I think we want a <trace>.export method rather than the current name, so some portion of this does need to be merged. I guess there is just some discussion about how to accomplish the addition of this method?

**Comment by ezyang** [comment]
I ported the TracingState binding code to pybind11. I just have to figure out how to make str(tracing_state) work... (defining a `__str__` method doesn't seem to be doing the right thing.)

**Comment by ezyang** [comment]
OK all done.

File: torch/csrc/jit/init.cpp [hunk 1]
```diff
@@ -39,42 +42,12 @@ using pass_type = void (std::shared_ptr<Graph>&);
 template<pass_type pass>
 PyObject * wrap_pass(PyObject *_unused, PyObject *py_state) {
   HANDLE_TH_ERRORS
-  THPUtils_assert(THPTracingState_Check(py_state), "expected a TracingState instance");
-  THPTracingState *state = (THPTracingState*)py_state;
-  pass(state->cdata->graph);
+  auto trace = py::handle(py_state).cast<std::shared_ptr<tracer::TracingState>>();
```
Comment (apaszke) [inline torch/csrc/jit/init.cpp]: It's enough to cast it to `tracer::TracingState*`. No need to create a new `shared_ptr`

File: torch/csrc/jit/python_ir.cpp [hunk 2]
```diff
@@ -214,11 +214,8 @@ void initPythonIRBindings(PyObject * module_) {
   .def_readonly("user",&Use::user)
   .def_readonly("offset",&Use::offset);
 
-  m.def("_jit_get_graph", [](py::object h_) {
-    auto py_state = h_.ptr();
-    JIT_ASSERTM(THPTracingState_Check(py_state), "expected a TracingState instance");
-    THPTracingState *state = (THPTracingState*)py_state;
-    return state->cdata->graph;
+  m.def("_jit_get_graph", [](std::shared_ptr<tracer::TracingState> s) {
```
Comment (apaszke) [inline torch/csrc/jit/python_ir.cpp]: Same here. You can take `tracer::TracingState&` as an argument

File: torch/csrc/jit/python_tracer.cpp [hunk 3]
```diff
@@ -12,169 +19,89 @@
...
+    });
+  // TODO: without defining traverse/clear, this leaks
+
+  m.def("_tracer_enter", [](std::vector<TraceInput> trace_inputs) {
+    return enter(std::move(trace_inputs));
+  });
+  m.def("_tracer_exit", [](variable_list var_outputs) {
+    tracer::exit(var_outputs);
+  });
+  m.def("_jit_createAutogradClosure", [](std::shared_ptr<TracingState> state) {
```
Comment (apaszke) [inline torch/csrc/jit/python_tracer.cpp]: `shared_ptr` again

File: torch/csrc/jit/python_tracer.cpp [hunk 4]
```diff
@@ -12,169 +19,89 @@
...
+      ss << *s.graph;
+      return ss.str();
+    })
+    .def("export", [](TracingState& s) {
+      return py::bytes(ExportGraph(s.graph, s.buffer_map, {}));
+    })
+    .def("export", [](TracingState& s, const std::vector<at::Tensor>& initializers) {
+      return py::bytes(ExportGraph(s.graph, s.buffer_map, initializers));
+    });
+  // TODO: without defining traverse/clear, this leaks
```
Comment (apaszke) [inline torch/csrc/jit/python_tracer.cpp]: How and why? It should be ok

**Comment by apaszke** [comment]
@zdevito yes we need a function like that, but it should not look liek this. I don't want to expose traces as part of the release, because they enter in a lot of subtle interactions with Variables and are not ready to be given to a large number of users. We should only advertize a `torch.toffee.export` function that takes in an `nn` model and saves it

**Pull request merged**